Gems/OpenXRVk/Code/Include/OpenXRVk/InputDeviceXRController.h

```diff
@@ -0,0 +1,452 @@
+
+        ////////////////////////////////////////////////////////////////////////////////////////////
+        InputDeviceXRController();
+        explicit InputDeviceXRController(const InputDeviceId& inputDeviceId,
+                                         ImplementationFactory implFactoryFn = &Implementation::Create);
+        AZ_DISABLE_COPY_MOVE(InputDeviceXRController);
+        ~InputDeviceXRController() override;
+
+        ////////////////////////////////////////////////////////////////////////////////////////////
+        // AzFramework::InputDevice interface
+        const InputChannelByIdMap& GetInputChannelsById() const override;
+        bool IsSupported() const override;
+        bool IsConnected() const override;
+
+        ////////////////////////////////////////////////////////////////////////////////////////////
+        // AzFramework::InputDeviceRequests interface
+        void TickInputDevice() override;
+
+        ////////////////////////////////////////////////////////////////////////////////////////////
+        // AzFramework::InputHapticFeedbackRequests interface
+        void SetVibration(float leftMotorSpeedNormalized, float rightMotorSpeedNormalized) override;
+
+        ////////////////////////////////////////////////////////////////////////////////////////////
+        //! Set the implementation of this input device
+        //! @param impl The Implementation to use
+        void SetImplementation(AZStd::unique_ptr<Implementation> impl);
+
+        ////////////////////////////////////////////////////////////////////////////////////////////
+        //! Set the implementation of this input device
+        //! @param implFactoryFn The Implementation factory create function to use
+        void SetImplementation(const ImplementationFactory& implFactoryFn);
+
+        ////////////////////////////////////////////////////////////////////////////////////////////
+        //! Get the non-owning pointer to the implementation of this input device
+        //! @return The raw implementation pointer
+        Implementation* GetImplementation() const;
+
+    protected:
+        ////////////////////////////////////////////////////////////////////////////////////////////
+        using ButtonChannelByIdMap = AZStd::unordered_map<InputChannelId, InputChannelDigital*>;
+        using TriggerChannelByIdMap = AZStd::unordered_map<InputChannelId, InputChannelAnalog*>;
+        using ThumbStickAxis1DChannelByIdMap = AZStd::unordered_map<InputChannelId, InputChannelAxis1D*>;
+        using ThumbStickAxis2DChannelByIdMap = AZStd::unordered_map<InputChannelId, InputChannelAxis2D*>;
+        using ThumbStickDirectionChannelByIdMap = AZStd::unordered_map<InputChannelId, InputChannelAnalog*>;
+        using ControllerAxis3DChannelByIdMap = AZStd::unordered_map<InputChannelId, InputChannelAxis3D*>;
+        using ControllerPoseChannelByIdMap = AZStd::unordered_map<InputChannelId, InputChannelQuaternion*>;
```

Comment:
The "one map to rule them all" does exist already, it's `m_allChannelsById`.  Since a lot of this code is following other InputDevice examples, I'll look at some of those and see what the main use of them is.  If we can remove all the "categorized" maps of channels, then that would be a fine change to make.

Reply:
It seems to simplify the code in `ProcessRawControllerState` having these categorized maps.  The reason is that these maps store a non-const pointer to the derived channel type, whereas the all-channels map stores a const pointer to the base channel.  So we would need to add a `const_cast` and another cast to call the non-const member function `ProcessRawInputEvent`.  I don't think it's worth changing now.

Reply:
Thanks for looking into the pros/cons of the change. I agree if the code that uses these members is going to get more complicated it's not worth changing.